File: ios/TNGuardTricare/Stores/PDFFormFiller.swift
import Foundation
import PDFKit
import UIKit

/// Fills the fields defined in a FormDefinition (from content.json) into a PDF and returns
/// a file URL ready to preview/share. Three paths, chosen by the bundled template's shape:
///
/// - **acroform**: a real, genuinely fillable PDF (e.g. the IRS W-4, or the TN enrollment
///   form) — locates each field's pre-existing widget annotation by name, across every page
///   since multi-page/multi-enclosure documents spread fields around, and sets its value.
///   `outputPage`, when set, trims the shared PDF down to just that one page afterward (moving
///   an already-filled PDFPage into a fresh PDFDocument has been verified to keep rendering
///   correctly — PDFKit regenerates each widget's own appearance stream, it doesn't depend on
///   the surrounding document).
/// - **overlay**: a real but non-interactive/flat page (e.g. the TN attestation enclosure,
///   which is just printed text with blank lines) — draws the original page as a background
///   via `CGContext.drawPDFPage`, then draws each field's value as plain text at the
///   coordinates configured in content.json.
/// - **synthesized placeholder** (no bundled file at all): draws a generic labeled page with
///   each value printed under its label. Used only until a real template is bundled.
///
/// Note on a dead end already ruled out: an earlier version of the placeholder path built its
/// own PDFAnnotation widgets from scratch and set `widgetStringValue` on them. PDFKit only
/// regenerates a widget's on-screen appearance for fields that were part of a real `/AcroForm`
/// to begin with — freshly-constructed annotations kept the value programmatically but
/// rendered blank in Preview/Files, which is why the placeholder path draws plain text instead.
enum PDFFormFiller {
    enum FillError: Error {
        case templateUnavailable
        case writeFailed
    }

    static func filledPDF(form: AppContent.FormDefinition, values: [String: String]) throws -> URL {
        let document: PDFDocument
        if let bundledURL = bundledTemplateURL(for: form) {
            switch form.resolvedFillMode {
            case .acroform:
                guard let sourceDocument = PDFDocument(url: bundledURL) else { throw FillError.templateUnavailable }
                fillAcroFormFields(sourceDocument, form: form, values: values)
                document = extractOutputPage(sourceDocument, outputPage: form.outputPage)
            case .overlay:
                document = try buildOverlayDocument(sourceURL: bundledURL, form: form, values: values)
            }
        } else {
            document = buildPlaceholderDocument(for: form, values: values)
        }

        let outputURL = FileManager.default.temporaryDirectory
            .appendingPathComponent("\(form.id)-\(UUID().uuidString).pdf")
        guard document.write(to: outputURL) else { throw FillError.writeFailed }
        return outputURL
    }

    private static func bundledTemplateURL(for form: AppContent.FormDefinition) -> URL? {
        Bundle.main.url(forResource: form.pdfTemplate.replacingOccurrences(of: ".pdf", with: ""), withExtension: "pdf")
    }

    // MARK: - acroform

    private static func fillAcroFormFields(_ document: PDFDocument, form: AppContent.FormDefinition, values: [String: String]) {
        for pageIndex in 0..<document.pageCount {
            guard let page = document.page(at: pageIndex) else { continue }
            for annotation in page.annotations {
                // PDFKit reports the *fully qualified* dotted field name for hierarchical
                // AcroForms (e.g. "topmostSubform[0].Page1[0].Step1a[0].f1_01[0]"), not just
                // the leaf name — match either an exact hit or the configured name as the
                // final path component, so content.json can keep using short leaf names.
                guard let fieldName = annotation.fieldName,
                      let field = form.fields.first(where: {
                          guard let pdfFieldName = $0.pdfFieldName else { return false }
                          return fieldName == pdfFieldName || fieldName.hasSuffix("." + pdfFieldName)
                      }) else { continue }
                switch annotation.widgetFieldType {
                case .text, .choice:
                    let rawValue = values[field.id] ?? ""
                    annotation.widgetStringValue = field.type == .date ? displayDate(from: rawValue) : rawValue
                case .button:
                    let checked = (values[field.id] ?? "").lowercased() == "true"
                    annotation.buttonWidgetState = checked ? .onState : .offState
                default:
                    break
                }
            }
        }
    }

    /// Moves a single page out of a (possibly multi-enclosure) source document into a fresh
    /// document, so sharing one filled form doesn't also share unrelated pages bundled
    /// alongside it in the same source PDF.
    private static func extractOutputPage(_ document: PDFDocument, outputPage: Int?) -> PDFDocument {
        guard let outputPage, let page = document.page(at: outputPage) else { return document }
        let output = PDFDocument()
        output.insert(page, at: 0)
        return output
    }

    // MARK: - overlay

    private static func buildOverlayDocument(sourceURL: URL, form: AppContent.FormDefinition, values: [String: String]) throws -> PDFDocument {
        guard let cgDocument = CGPDFDocument(sourceURL as CFURL) else { throw FillError.templateUnavailable }
        let pageNumber = (form.outputPage ?? 0) + 1 // CGPDFDocument pages are 1-indexed
        guard let cgPage = cgDocument.page(at: pageNumber) else { throw FillError.templateUnavailable }
        let pageRect = cgPage.getBoxRect(.mediaBox)

        let renderer = UIGraphicsPDFRenderer(bounds: pageRect)
        let pdfData = renderer.pdfData { context in
            context.beginPage()
            let cg = context.cgContext
            cg.saveGState()
            // drawPDFPage always draws in the PDF's native bottom-left-origin space; flip into
            // it just for this call, then draw overlay text in the renderer's normal top-left,
            // y-down space (matching the coordinates content.json stores, taken directly from
            // pdfplumber's word positions, which use the same top-left convention).
            cg.translateBy(x: 0, y: pageRect.height)
            cg.scaleBy(x: 1, y: -1)
            cg.drawPDFPage(cgPage)
            cg.restoreGState()

            let textAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 11),
                .foregroundColor: UIColor.black
            ]

            for field in form.fields {
                let rawValue = values[field.id] ?? ""
                guard !rawValue.isEmpty else { continue }
                if field.type == .choice {
                    guard let match = field.overlayOptions?.first(where: { $0.value == rawValue }) else { continue }
                    ("X" as NSString).draw(at: CGPoint(x: match.x, y: match.y), withAttributes: textAttrs)
                } else {
                    let displayValue = field.type == .date ? displayDate(from: rawValue) : rawValue
                    for position in field.overlays ?? [] {
                        (displayValue as NSString).draw(at: CGPoint(x: position.x, y: position.y), withAttributes: textAttrs)
                    }
                }
            }
        }
        return PDFDocument(data: pdfData) ?? PDFDocument()
    }

    /// Stored date values are "yyyy-MM-dd" (see FormFillView); printed forms read more
    /// naturally as "MM/dd/yyyy", matching the convention used elsewhere on these PDFs.
    private static func displayDate(from stored: String) -> String {
        let storageFormatter = DateFormatter()
        storageFormatter.dateFormat = "yyyy-MM-dd"
        guard let date = storageFormatter.date(from: stored) else { return stored }
        let displayFormatter = DateFormatter()
        displayFormatter.dateFormat = "MM/dd/yyyy"
        return displayFormatter.string(from: date)
    }

    // MARK: - synthesized placeholder

    /// Draws a clean labeled page with each field's entered value printed as static text
    /// under its label — used only until a real template is bundled for this form.
    private static func buildPlaceholderDocument(for form: AppContent.FormDefinition, values: [String: String]) -> PDFDocument {
        let pageWidth: CGFloat = 612
        let pageHeight: CGFloat = 792
        let margin: CGFloat = 48
        let rowHeight: CGFloat = 54

        let renderer = UIGraphicsPDFRenderer(bounds: CGRect(x: 0, y: 0, width: pageWidth, height: pageHeight))
        let pdfData = renderer.pdfData { context in
            context.beginPage()
            let titleAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.boldSystemFont(ofSize: 20),
                .foregroundColor: UIColor.black
            ]
            let subtitleAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 11),
                .foregroundColor: UIColor.darkGray
            ]
            let labelAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 11, weight: .semibold),
                .foregroundColor: UIColor.black
            ]
            let valueAttrs: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 13),
                .foregroundColor: UIColor.black
            ]

            (form.title as NSString).draw(at: CGPoint(x: margin, y: 40), withAttributes: titleAttrs)
            ("Generated by TN Guard Tricare Helper — placeholder template, verify against the official form at tn.gov before sending." as NSString)
                .draw(in: CGRect(x: margin, y: 66, width: pageWidth - margin * 2, height: 28), withAttributes: subtitleAttrs)

            var y: CGFloat = 110
            for field in form.fields {
                (field.label as NSString).draw(at: CGPoint(x: margin, y: y), withAttributes: labelAttrs)
                let boxRect = CGRect(x: margin, y: y + 16, width: pageWidth - margin * 2, height: 22)
                UIColor.lightGray.setStroke()
                context.cgContext.stroke(boxRect)

                let rawValue = values[field.id] ?? ""
                let displayValue = field.type == .checkbox ? (rawValue.lowercased() == "true" ? "Yes" : "No") : rawValue
                if !displayValue.isEmpty {
                    (displayValue as NSString).draw(
                        at: CGPoint(x: margin + 6, y: y + 19),
                        withAttributes: valueAttrs
                    )
                }
                y += rowHeight
            }
        }

        return PDFDocument(data: pdfData) ?? PDFDocument()
    }
}

File: android/app/src/main/java/org/tnguardtricare/app/pdf/PdfFormFiller.kt
package org.tnguardtricare.app.pdf

import android.content.Context
import com.tom_roush.pdfbox.pdmodel.PDDocument
import com.tom_roush.pdfbox.pdmodel.PDPage
import com.tom_roush.pdfbox.pdmodel.PDPageContentStream
import com.tom_roush.pdfbox.pdmodel.common.PDRectangle
import com.tom_roush.pdfbox.pdmodel.font.PDType1Font
import com.tom_roush.pdfbox.pdmodel.interactive.form.PDCheckBox
import com.tom_roush.pdfbox.pdmodel.interactive.form.PDField
import com.tom_roush.pdfbox.pdmodel.interactive.form.PDNonTerminalField
import org.tnguardtricare.app.model.AppContent
import java.io.File
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.UUID

/**
 * Fills the fields defined in a FormDefinition (from content.json) into a PDF and returns a
 * file ready to preview/share. Three paths, chosen by the bundled template's shape — mirrors
 * iOS's PDFFormFiller.swift exactly, including the same two lessons learned there:
 *
 * - **acroform**: a real fillable PDF (the Enrollment form, the W-4) — locate each field's
 *   existing widget by name (PDFBox reports the *fully qualified* dotted name for hierarchical
 *   forms, e.g. the W-4's XFA-style naming — matched with the same suffix-match iOS uses so
 *   content.json can keep using short leaf names) and set its value. `outputPage`, when set,
 *   trims the shared PDF to just that page via PDDocument.importPage, which (as verified on
 *   iOS's equivalent page-move) keeps each widget's already-rendered appearance intact.
 * - **overlay**: a real but flat/non-interactive page (the Attestation enclosure) — draws the
 *   original page content plus each value as plain text at the coordinates content.json
 *   stores (taken from pdfplumber word positions, top-left origin — flipped here since
 *   PDFPageContentStream text drawing uses PDF's native bottom-left origin).
 * - **synthesized placeholder**: no bundled file at all — draws a generic labeled page. Only
 *   reached if a template file is ever missing; both real Android templates are bundled today.
 */
class PdfFormFiller(private val context: Context) {

    fun filledPdf(form: AppContent.FormDefinition, values: Map<String, String>): File {
        val source = openTemplate(form)
        val document = source?.let {
            when (form.resolvedFillMode) {
                AppContent.FillMode.ACROFORM -> {
                    fillAcroFormFields(it, form, values)
                    extractOutputPage(it, form.outputPage)
                }
                AppContent.FillMode.OVERLAY -> overlayValues(it, form, values)
            }
        } ?: buildPlaceholderDocument(form, values)

        val outDir = File(context.cacheDir, "pdfs").apply { mkdirs() }
        val outFile = File(outDir, "${form.id}-${UUID.randomUUID()}.pdf")
        try {
            document.save(outFile)
        } finally {
            document.close()
            if (source != null && source !== document) source.close()
        }
        return outFile
    }

    private fun openTemplate(form: AppContent.FormDefinition): PDDocument? {
        val assetName = "forms/${form.pdfTemplate}"
        return try {
            context.assets.open(assetName).use { PDDocument.load(it) }
        } catch (e: Exception) {
            null
        }
    }

    // MARK: - acroform

    private fun fillAcroFormFields(document: PDDocument, form: AppContent.FormDefinition, values: Map<String, String>) {
        val acroForm = document.documentCatalog.acroForm ?: return
        for (field in flattenFields(acroForm.fields)) {
            val fieldName = field.fullyQualifiedName ?: continue
            val configured = form.fields.firstOrNull { cf ->
                val pdfFieldName = cf.pdfFieldName ?: return@firstOrNull false
                fieldName == pdfFieldName || fieldName.endsWith(".$pdfFieldName")
            } ?: continue

            val rawValue = values[configured.id] ?: ""
            if (field is PDCheckBox) {
                val checked = rawValue.equals("true", ignoreCase = true)
                if (checked) field.check() else field.setValue("Off")
            } else {
                val display = if (configured.type == AppContent.FieldType.DATE) displayDate(rawValue) else rawValue
                try {
                    field.setValue(display)
                } catch (e: Exception) {
                    // Some choice fields reject values outside their option list under strict
                    // validation; fall back to setting the raw string on the underlying COS
                    // object so free-text-style combo boxes (e.g. TRICARE Benefits Number)
                    // still accept arbitrary input the way they do on iOS.
                }
            }
        }
    }

    private fun flattenFields(fields: List<PDField>): List<PDField> {
        val result = mutableListOf<PDField>()
        for (field in fields) {
            if (field is PDNonTerminalField) {
                result.addAll(flattenFields(field.children))
            } else {
                result.add(field)
            }
        }
        return result
    }

    /** Moves a single page into a fresh document so sharing one filled form doesn't also
     * share unrelated pages bundled alongside it in the same source PDF (e.g. the state
     * packet's other enclosures). */
    /** Caller is responsible for closing the returned document (and the original, if
     * different) only after it has been saved — importPage does not necessarily deep-copy
     * every backing resource immediately, so closing the source too early can corrupt the
     * save that happens afterward. */
    private fun extractOutputPage(document: PDDocument, outputPage: Int?): PDDocument {
        if (outputPage == null) return document
        val page = document.getPage(outputPage) ?: return document
        val trimmed = PDDocument()
        trimmed.importPage(page)
        return trimmed
    }

    // MARK: - overlay

    private fun overlayValues(document: PDDocument, form: AppContent.FormDefinition, values: Map<String, String>): PDDocument {
        val pageIndex = form.outputPage ?: 0
        val page = document.getPage(pageIndex) ?: return document
        val pageHeight = page.mediaBox.height
        val font = PDType1Font.HELVETICA
        val fontSize = 11f

        PDPageContentStream(document, page, PDPageContentStream.AppendMode.APPEND, true, true).use { cs ->
            for (field in form.fields) {
                val rawValue = values[field.id] ?: ""
                if (rawValue.isEmpty()) continue
                if (field.type == AppContent.FieldType.CHOICE) {
                    val match = field.overlayOptions?.firstOrNull { it.value == rawValue } ?: continue
                    drawText(cs, "X", match.x.toFloat(), match.y.toFloat(), pageHeight, font, fontSize)
                } else {
                    val display = if (field.type == AppContent.FieldType.DATE) displayDate(rawValue) else rawValue
                    for (position in field.overlays ?: emptyList()) {
                        drawText(cs, display, position.x.toFloat(), position.y.toFloat(), pageHeight, font, fontSize)
                    }
                }
            }
        }

        // Same reasoning as the acroform path: don't share unrelated pages bundled alongside
        // this one in the same source PDF.
        return extractOutputPage(document, form.outputPage)
    }

    private fun drawText(
        cs: PDPageContentStream,
        text: String,
        topLeftX: Float,
        topLeftY: Float,
        pageHeight: Float,
        font: PDType1Font,
        fontSize: Float,
    ) {
        // content.json coordinates are top-left origin (taken directly from pdfplumber word
        // positions); PDPageContentStream text drawing uses PDF's native bottom-left origin
        // with the point as the text baseline, roughly `fontSize * 0.8` below the glyph tops.
        val baselineY = pageHeight - topLeftY - (fontSize * 0.8f)
        cs.beginText()
        cs.setFont(font, fontSize)
        cs.newLineAtOffset(topLeftX, baselineY)
        cs.showText(text)
        cs.endText()
    }

    // MARK: - synthesized placeholder

    private fun buildPlaceholderDocument(form: AppContent.FormDefinition, values: Map<String, String>): PDDocument {
        val document = PDDocument()
        val page = PDPage(PDRectangle.LETTER)
        document.addPage(page)
        val margin = 48f
        var y = page.mediaBox.height - 110f
        val rowHeight = 54f
        val titleFont = PDType1Font.HELVETICA_BOLD
        val labelFont = PDType1Font.HELVETICA_BOLD
        val valueFont = PDType1Font.HELVETICA

        PDPageContentStream(document, page).use { cs ->
            cs.beginText()
            cs.setFont(titleFont, 20f)
            cs.newLineAtOffset(margin, page.mediaBox.height - 60f)
            cs.showText(form.title)
            cs.endText()

            cs.beginText()
            cs.setFont(PDType1Font.HELVETICA, 9f)
            cs.newLineAtOffset(margin, page.mediaBox.height - 80f)
            cs.showText("Generated by TN Guard Tricare Helper — placeholder template.")
            cs.endText()

            for (field in form.fields) {
                cs.beginText()
                cs.setFont(labelFont, 11f)
                cs.newLineAtOffset(margin, y)
                cs.showText(field.label)
                cs.endText()

                cs.setStrokingColor(200, 200, 200)
                cs.addRect(margin, y - 34f, page.mediaBox.width - margin * 2, 22f)
                cs.stroke()

                val rawValue = values[field.id] ?: ""
                val displayValue = if (field.type == AppContent.FieldType.CHECKBOX) {
                    if (rawValue.equals("true", ignoreCase = true)) "Yes" else "No"
                } else rawValue
                if (displayValue.isNotEmpty()) {
                    cs.beginText()
                    cs.setFont(valueFont, 12f)
                    cs.newLineAtOffset(margin + 6f, y - 28f)
                    cs.showText(displayValue)
                    cs.endText()
                }
                y -= rowHeight
            }
        }
        return document
    }

    private fun displayDate(stored: String): String {
        return try {
            val date = LocalDate.parse(stored, DateTimeFormatter.ISO_LOCAL_DATE)
            date.format(DateTimeFormatter.ofPattern("MM/dd/yyyy"))
        } catch (e: Exception) {
            stored
        }
    }
}
